calculator/main.py

```diff
@@ -26,7 +31,22 @@ def perform_operation(calc, operation, params):
     :param params: Tuple containing the list of nums to operate with.
                    ie: (1, 2, 3, 4.5, -2)
     """
-    pass
+    if operation not in calc['operations'].keys():
+        raise InvalidOperation("{} operation not supported.".format(operation))
+    elif params == False:
+        raise InvalidParams("Given params are invalid.")
+    else: 
+        for i in params:
+            if isinstance(i, bool) or isinstance(i, str):
+                raise InvalidParams("Given params are invalid.")
+        
+    
+    result = calc['operations'][operation](*params)
+    now = datetime.now()
+    time_list = [now.strftime('%Y-%m-%d %H:%M:%S'), operation, params, result]
+    timestamp = tuple(time_list)
+    calc['history'].append(timestamp)
```

Comment:
Good call, when we initially tried to implement, we got an invalid tuple error message and tried to get the test to pass.  Need to go back and refactor.